proposals/0001-dependency-injection.md

```diff
@@ -0,0 +1,642 @@
+classpath than the application module `configureApplication`, which has no knowledge of the implementation details.
+
+In the next two sections, we'll provide the API design for both declaring and resolving dependencies.
+
+## Declaration
+[declaration]: #declaration
+
+As with declaring extensions in a Ktor application, users can expect to declare their dependencies from the
+same scope: `Application`.
+
+Here is how to declare dependencies from an application module:
+
+```kotlin
+fun Application.declareDependencies() {
+    dependencies {
+        // Base case
+        provide<DataSource> { PostgresDataSource("jdbc:postgresql://localhost:5440/test") }
+        // With named instance, using resolve() from the provider context
+        provide<DataSource>("mongo") { MongoDataSource(resolve()) }
+        // Using constructor injection from a class reference
+        provide<Repository<Message>>(MessageRepository::class)
+        // Using resolution and constructor injection from the provider context
+        provide { RoomRepository(resolve<DataSource>(), create(::RoomAuthorityAdapter)) }
+    }
+}
+```
+
+### Naming
+[naming]: #naming
+
+For the *dependencies* scope:
+
+| Alternative | Reason for avoiding                  |
+|-------------|--------------------------------------|
+| di          | Not immediately recognizable acronym |
+| inject      | confusing for Koin users             |
+| providers   | too abstract                         |
+
+For the *provide* function:
+
+| Alternative | Reason for avoiding                                    |
+|-------------|--------------------------------------------------------|
+| inject      | confusing for Koin users                               |
+| supply      | this suggests that there is some scarcity of instances |
+| declare     | too abstract and non-operative                         |
+| single      | awkward wording out of context                         |
+
+For the *create* function:
+
+| Alternative | Reason for avoiding                                          |
+|-------------|--------------------------------------------------------------|
+| construct   | too specific; this call should allow non-constructor lambdas |
+| inject      | not the most important detail of the function                |
+
+
+## Resolution
+[resolution]: #resolution
+
+In practice, we can expect the declaring modules to be few, and the resolving modules to be many.
+
+Here are the examples of how to resolve dependencies from a Ktor module:
+
+```kotlin
+fun Application.configureRouting() {
+    // Resolve by property delegation
+    val users: Repository<User> by dependencies
+    // Using a named instance
+    val mongo: DataSource by dependencies.named("mongo")
+    // Resolve from the dependencies property in the Application scope
+    val messages: Repository<Message> = dependencies.resolve()
```

Comment:
Should we mention here that dependencies are instantiated lazily when delegate is used and eagerly when `resolve()` is used?